### ISSUE TITLE
sbc-Inbound reports timeout

### ISSUE THREAD

**Issue by zvoneos** [jambonz/sbc-inbound#23]
We have deployed jambonz on GCP Kubernetes. We wanted to do a small load and stress testing, but we noticed that even on small load after some time (we are simulating calls using SIPP) SBC-Inbound starts to report errors like this:

```
"jsonPayload": {
      "pid": 20,
      "msg": "unexpected error routing inbound call",
      "err": {
        "message": "rtpengine timeout",
        "type": "RtpEngineError",
        "stack": "Error: rtpengine timeout\n    at Client._onMessageTimeout (/opt/app/node_modules/@jambonz/rtpengine-utils/node_modules/rtpengine-client/lib/BaseClient.js:122:14)\n    at listOnTimeout (node:internal/timers:559:17)\n    at processTimers (node:internal/timers:502:7)"
      },
      "callId": "208-769036@10.60.21.2",
      "hostname": "sbc-inbound-869bff9b45-hrz8g",
      "level": 50
    },
```

It seems as if after some time of running our test calls, response (200 OK) that is sent by the feature-server gets wrongly handled. 

By looking at the feature-server logs I see that it accepts the INVITE properly, triggers the application, and handles application response properly (answer&play) and feature server sends 200 OK to the sbc-inbound. 

In normal calls, 200 OK is sent within miliseconds after the INVITE and is visible in the sbc-inbound trace. 

```
"insertId": "cykwwiqq92jtyaef",
    "jsonPayload": {
      "level": 30,
      "callId": "200-769036@10.60.21.2",
      "pid": 20,
      "msg": "call connected successfully to feature server",
      "hostname": "sbc-inbound-869bff9b45-hrz8g"
    },
```

But, in case of error this message is not visible in the sbc-inbound trace and after 3 sec of the INVITE we see that error.

When we collect rtp-trace on feature-server and on sbc-inbound we see that feature-server has sent and sbc-inbound has received 200 OK on time, but it seems as if it somehow got lost.

The CSP is very low and memory and CPU usage on all nodes is minimal.

**Comment by davehorton** [comment]
hmm, interesting it seems like sbc-inbound receives the 200 OK from the FS, sends the 'answer' command to rtpengine but times out waiting for a response from rtp engine.  

Could you share the sipp script?  I can then try to recreate on my cluster

**Comment by davehorton** [comment]
I have done similar load testing without issues, so would like to get to the bottom of this

**Comment by davehorton** [comment]
also might be good to check the rtpengine logs to see if any errors appear there

**Comment by davehorton** [comment]
you should also see this log message from sbc-inbound after startup, and any time the rtpengine nodes scale out or in:
```
rtpserver endpoints have been updated
```
and it will list the pod IPs of the rtpengine server instances it will be sending to.  Then, sbc-inbound sends commands to rtpengine at those Ips, over udp to port 22222/udp.  

Its quite strange, because from what you report the 'offer' command was sent to rtpengine and worked, but the 'answer' command sent to the same rtpengine times out

**Comment by zvoneos** [comment]
SIPP Script we have is very simple. We just send INVITE to number 300 (behind which on Jambonz we have application that just answers the call and some mp3 file), we expect 200 and then we play from our side some wav file and after 300 sec we hangup.

When traffic is very low we run it constantly, but as soon as we increase it to 1.75 CPS (with duration of 30 sec it is around 50 concurrent calls) it starts to fail.

Just to highlight (don't know was it obvious from the originalpost), we see this message `"call connected successfully to feature server"` when call is properly handled. We don't see it for failed calls.

**Comment by davehorton** [comment]
OK, I am running that test against my cluster, currently running 2 CPS with 60-90 concurrent calls, do not see the failure yet.  I will keep testing, but one thing I noticed is that my helm chart was setting some debug levels quite high -- particularly for rtpengine which was logging every rtp packet.  You should probably apply my changes from this commit which turns them to a more appropriate level for a production system.

https://github.com/jambonz/helm-charts/commit/9a2b496278e0b73eaa8e92e243a3b39645eb74eb

I am testing on a 3-node cluster with e2-medium instances, per below.  What does your setup look like?

![image](https://user-images.githubusercontent.com/1884973/155171925-353dbd86-283c-4b9c-afe8-fcaf2c794473.png)

**Comment by davehorton** [comment]
I've done quite a bit of load testing today, and have not recreated that specific error.  Can you give details of instance types and  number of nodes you are using, and also re-run your tests after turning down those logs?

Meanwhile, here are my results.  

I ran a first round of tests, starting with the cluster as configured above -- 3 nodes, one in each node pool, each node running on a e2-medium instance ytpe.

|CPS| max concurrent calls|number of nodes in default node pool|result|
|-----|---------------------|-------------------------------------|---|
|2|90|1|no failures|
|5|150|3|no failures|
|7|200|experienced 503 call set up failures|

Now, the e2-medium instance type is not an instance I would use for a production VoIP system.  It has only one core and not much memory.  For production, I recommend using one of the compute-optimized instance.  So next, I configured what I would consider to be more of a production cluster: I used the c2d-standard-2 instance which is a compute-optimized instance with 2vCPUs and 8 GB of memory.  This is a more suitable instance for running the feature server and rtp deployments on.  I started with a 3-node cluster again, one node in each of the 3 node pools (sip, rtp, default), then scaled up the default node pool and feature server replicas as I went.

Results were as follows:

|CPS| max concurrent calls|number of nodes in default node pool|result|
|-----|---------------------|-------------------------------------|---|
|10|300|1|no failures|
|15|500|2|no failures|
|20|600|2|no failures|
|30|750|3|no failures|
|50|1000|3|503 call setup failures|

I could have continued scaling out the default node pool and the number of feature server replicas beyond that if needed to reach higher levels, but I have not yet been able to recreate that specific error message you are seeing in your testing.

I used your sipp scenario, with a [jambonz app](https://gist.github.com/davehorton/1bdc7283aed77d13d3de5394628b27c8) that simply answers and plays sythesized speech until the caller hangs up.

**Comment by zvoneos** [comment]
Hello Dave. We have adjusted the log level as mentioned and indeed it resolved the issue. Thank you for the support and sorry for not noticing it at first. 

Our initial nodes that we used for testing are minimal, but since we did not have any issues with CPU and memory they were sufficient for that initial load.

We will now try to do a load test with mixed traffic (different call scenarios: answered VoIP calls, rejected VoIP calls, play etc.) and will share also our results.

Thank you for your support!
